coverage-dump: Resolve global file IDs to filenames

Zalathar opened the pull request:
The coverage-dump tool, used by coverage tests, currently includes “global file ID” numbers in its dump output.

This PR adds support for parsing coverage filename information from LLVM assembly `.ll` files, and resolving those file IDs to the corresponding filename, for inclusion in dump output.

This makes dump output more informative, especially for test cases involving multiple files, and will be important for testing expansion region support in the future.

---

The bootstrap changes don't necessarily have to land at the same time (e.g. they could be deferred to after the stage0 redesign if requested), but I would prefer to land them now if possible.

Zalathar committed:
bootstrap: Add check/test/run steps for src/tools/coverage-dump

This also causes the coverage-dump unit tests to run in CI and `./x test` by
default.

Zalathar committed:
coverage-dump: Include filenames hash in covfun line data

Zalathar committed:
coverage-dump: Extract some common code to an `llvm_utils` submodule

Zalathar committed:
coverage-dump: Extract a common parser method for maybe-compressed bytes

Zalathar committed:
coverage-dump: Make filenames available to covfun record dumping

Actually printing the filenames is deferred to a subsequent commit that will
simultaneously bless all affected tests.

Zalathar committed:
coverage-dump: Dump filenames instead of global file IDs (and bless)

clubby789 commented:
The bootstrap step additions look reasonable to me

clubby789 commented on src/bootstrap/src/core/build_steps/check.rs:
```diff
@@ -527,3 +527,59 @@ tool_check_step!(Bootstrap { path: "src/bootstrap", default: false });
 tool_check_step!(RunMakeSupport { path: "src/tools/run-make-support", default: false });
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct CoverageDump {}
```
nit: Could just be a unit struct

```suggestion
pub(crate) struct CoverageDump;
```

clubby789 commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -54,6 +54,7 @@ impl Step for CrateBootstrap {
         run.path("src/tools/jsondoclint")
             .path("src/tools/suggest-tests")
             .path("src/tools/replace-version-placeholder")
+            .path(check::CoverageDump::PATH)
```
nit: I'm not sure about putting the paths and display names in constants, since it's inconsistent with the other entries here and makes it a little hard to discover at a glance.
Up to you if the advantages here outweigh the downsides

Zalathar replied:
It seemed like a good idea at the time, but I think I'll go back to hard-coded paths, since there isn't a big practical advantage to using a constant here.

Zalathar commented:
Rebased, then addressed review feedback, and added some relevant FIXMEs and other comments [(diff)](https://github.com/rust-lang/rust/compare/aca0f349df6e8f842e50b797252f6a1273937512..c65fc17747c1ea92093550c46449f97859b46ec8).

Zalathar commented:
(Rebased; no changes.)

clubby789 commented on src/tools/coverage-dump/src/llvm_junk.rs:
nit: I think `llvm_junk` is a bit confusing; maybe `utils`/`llvm_utils`?

clubby789 reviewed (approved):
Changes LGTM aside from a small nit, r=me with or without it

Zalathar commented:
Rebased, and renamed `llvm_junk` to `llvm_utils` [(diff)](https://github.com/rust-lang/rust/compare/dfc04ed396ae331a477d6935b7d9aa3283e617e9..c53a76743c01afa356ae5301bc1b35044354da7a).

@bors r=clubby789 rollup

bors commented:
:pushpin: Commit c53a76743c01afa356ae5301bc1b35044354da7a has been approved by `clubby789`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=clubby789 c53a76743c01afa356ae5301bc1b35044354da7a -->
<!-- homu: {"type":"Approved","sha":"c53a76743c01afa356ae5301bc1b35044354da7a","approver":"clubby789","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
